### ISSUE TITLE
NavigationPage BackButtonPressed sent ignores Android OnBackPressed

### ISSUE THREAD

**Issue by dansiegel** [dotnet/maui#15296]
### Description

Android has a global handler for when the Hardware back button is pressed. This is separate from the visual back in the NavigationBar of a NavigationPage. Currently `NavigationPage.OnBackButtonPressed()` is executed prior to the Android Lifecycle Event handler for `OnBackPressed`. It also has no arguments passed to know whether the event is related to the NavigationBar or a device back button. 

As a result there is no way to prevent a NavigationPage from executing it's back and then overriding the back behavior from the Lifecycle Event handler.

### Steps to Reproduce

1. Create an app
2. Create a custom NavigationPage. Override OnBackButtonPressed so you can see when this is invoke.
3. Create two sample pages `ViewA` & `ViewB`
4. Push ViewA and ViewB in the NavigationPage
5. Set the NavigationPage as the MainPage in the Application
6. Launch app on Android
7. Press the hardware back button on the Android device/emulator

### Link to public reproduction project repository

https://github.com/dansiegel/MauiBackBug

### Version with bug

7.0.86

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

All android versions

**Comment by VitalyKnyazev** [comment]
Good that I have found this issue - was about to raise a new one.

It looks like all lifecycle.AddAndroid+android.OnBackPressed handlers **in any module** of MAUI app are called one by one and this chain cannot be interrupted by returning true from android.OnBackPressed (I believe a better mechanism is needed anyway, see below). 

In my case it is Mopups+Prism+MAUI. I would expect that Mopups handles first Back button click and then everything stops.
But in fact we have this:

- MAUI handler is called and page navigates back (MAUI handler is always called first as it is registered in MauiAppBuilder constructor, order of these calls does not matter much ConfigureMopups, UsePrism, UseMauiApp);
- Mopups handler is called and popup is about to be closed (it is async);
- Prism handler is called and page navigates back one more time (or we go to home screen of Android emulator);
- quite often it all just crashes or freezes.

Above behaviour seems to be wrong and cannot be fixed without patching all 3 modules, and overall looks like a showstopper for now.

**Comment by SkytAsul** [comment]
Any news on this? This is a serious drawback that prevents the use of MAUI in a production app.

**Comment by XamlTest** [comment]
Verified this on Visual Studio Enterprise 17.7.0 Preview 3.0. Repro on Android 13.0-API33 .NET 8 with below Project:
[MauiBackBug.zip](https://github.com/dotnet/maui/files/12078116/MauiBackBug.zip)

**Comment by thisisthekap** [comment]
@jfversluis This is a blocker migrating to MAUI.

**Comment by jfversluis** [comment]
I'm not entirely sure if I understand what the issue is here and maybe even more: this is then a difference comparing to Xamarin.Forms? How did this behave in Forms and how is that different from .NET MAUI?

In the application output I notice a Android warning as well that suggests adding `android:enableOnBackInvokedCallback="true"` to your `AndroidManifest.xml`, that does seem to change the behavior, does that seem to do what you expect more?

**Comment by albilaga** [comment]
@jfversluis in Xamarin.Forms when user click back button it triggers `OnBackButtonPressed`, but in MAUI it doesn't. You can check with this branch https://github.com/albilaga/MauiPlayground/tree/back_button. As you can see I already added `android:enableOnBackInvokedCallback="true"` in [here](https://github.com/albilaga/MAUIPlayground/blob/back_button/MAUIPlayground/Platforms/Android/AndroidManifest.xml#L3) 

But whenever user hit back button hardware in android, [this line](https://github.com/albilaga/MAUIPlayground/blob/back_button/MAUIPlayground/SecondPage.xaml.cs#L12) is never triggered

**Comment by thisisthekap** [comment]
@jfversluis Can you think of a workaround on that one, or is MAUI adapted to behave like we were used to in Xamarin.Forms?

**Comment by s4jc** [comment]
if you want to interrupt either back navigation (hardware back or nav bar), you can override OnNavigating, and check the ShellNavigatingEventArgs.Source to see if it is a ShellNavigationSource.Pop or not.

**Comment by albilaga** [comment]
@s4jc we are not using shell. is that still working even when we are not using shell?

**Comment by RandomMauiDev** [comment]
This is happening to me as well, I am using Mopups to display a dialog box but when I press Android hardware back button both popup and page is popped. This is a blocker for me as well..

**Comment by s4jc** [comment]
@albilaga Sorry, I think you're right, my proposal will only work for a shell application.  Now that I think about it, we might have switched to the shell application approach *specifically* because of this problem.  (We absolutely needed to interrupt the back navigation to do data validations)

**Comment by henda79** [comment]
I've just came up against this issue when trying to stop users closing pages they shouldn't. Since I am using Prism I was able to workaround using this code.

```
    private bool _backPressed;

    protected override bool OnBackButtonPressed()
    { 
        _backPressed = true;
        return true;
    }

    public bool CanNavigate(INavigationParameters parameters)
    {
        if (!_backPressed)
        {
            return true;
        }

        _backPressed = false;
        return false;
    }
```

**Comment by maonaoda** [comment]
This is a serious hindrance

**Comment by jaisontjohn-gss** [comment]
> android:enableOnBackInvokedCallback="true"

Removing this makes the backbuttonpressed work actually.

**Comment by thomasgalliker** [comment]
Setting `android:enableOnBackInvokedCallback="true"` in your AndroidManifest.xml changes the behavior of the Hardware-Go-Back button. It moves the app to the background - no matter on what page you currently are.

What we had in the past (with Xamarin) was the method `OnBackButtonPressed ` which could be used to intercept and prevent back-navigation. You can still override OnBackButtonPressed but it doesn't do anything. It's not even hit when I set a breakpoint.

This is very critical for us to migrate from Xamarin to MAUI. We have several pages which MUST NOT be navigated back.
Anyone from the MAUI core team, @jfversluis or @samhouts, do you see a solution for this issue in the near future?

**Comment by maonaoda** [comment]
@thomasgalliker are you using the Prism in maui?

**Comment by thomasgalliker** [comment]
Okay guys, I owe you a coffee/beer. My fault. The debugger did not hit the `OnBackButtonPressed` (because... Rider already left for the weekend obviously, idk). 
@jfversluis I use MauiVersion 8.0.80, updated all workloads, reinstalled Rider - and it seems to work now (reproduced with this [simple navigation sample app](https://github.com/hsrcasnet/MauiSamples/tree/main/NavigationDemo)). Sorry for wasting your time - and thanks btw. for all the efforts!!
@maonaoda Thanks for the hint. I saw that Prism is intercepting `OnBackButtonPressed` - so, in the Prism app I had to find a workaround to catch and stop backward navigation on Android (`IConfirmNavigation`).
